pkg/config/interface.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"sigs.k8s.io/kubebuilder/v3/pkg/model/resource"
+)
+
+// Config defines the interface that project configuration types must follow
+type Config interface {
+	/* Version */
+
+	// GetVersion returns the current project version
+	GetVersion() Version
+
+	/* String fields */
+
+	// GetDomain returns the project domain
+	GetDomain() string
+	// SetDomain sets the project domain
+	SetDomain(domain string) error
+
+	// GetRepository returns the project repository.
+	GetRepository() string
+	// SetRepository sets the project repository
+	SetRepository(repository string) error
+
+	// GetName returns the project name
+	// This method was introduced in project version 3-alpha.
+	GetName() string
+	// SetName sets the project name
+	// This method was introduced in project version 3-alpha.
+	SetName(name string) error
+
+	// GetLayout returns the config layout
+	// This method was introduced in project version 3-alpha.
+	GetLayout() string
+	// SetLayout sets the Config layout
+	// This method was introduced in project version 3-alpha.
+	SetLayout(layout string) error
+
+	/* Boolean fields */
+
+	// IsMultiGroup checks if multi-group is enabled
+	IsMultiGroup() bool
+	// SetMultiGroup enables multi-group
+	SetMultiGroup() error
+	// ClearMultiGroup disables multi-group
+	ClearMultiGroup() error
+
+	// IsComponentConfig checks if component config is enabled
+	// This method was introduced in project version 3-alpha.
+	IsComponentConfig() bool
+	// SetComponentConfig enables component config
+	// This method was introduced in project version 3-alpha.
+	SetComponentConfig() error
+	// ClearComponentConfig disables component config
+	// This method was introduced in project version 3-alpha.
+	ClearComponentConfig() error
+
+	/* Resources */
+
+	// ResourcesLength returns the number of tracked resources
+	ResourcesLength() int
+	// HasResource checks if the provided GVK is stored in the Config
+	HasResource(gvk resource.GVK) bool
+	// GetResource returns the stored resource matching the provided GVK
+	GetResource(gvk resource.GVK) (resource.Resource, error)
+	// AddResource adds the provided resource if it was not present, no-op if it was already present
+	AddResource(res resource.Resource) error
+	// UpdateResource adds the provided resource if it was not present, modifies it if it was already present
+	UpdateResource(res resource.Resource) error
```

Comment:
It is missing getResources to return all resources from the config.

Reply:
As slices and underlying arrays are mutable, we can't just return the field. We will need to (deep) copy the whole array. This also includes making a copy of anything that is a pointer. This is a very expensive operation.

Which is the usage of `GetResources`?

Reply:
Hi @Adirio, 

We need to be able to GET all resources from a config. If this layout does not allow it then we cannot move forward with this proposal. 

Otherwise, we cannot move forward with this proposal. The KB is used and extends for we are to create other plugins and customize them. in this way, has many scenarios that we need to GET all resources from the config.